sdks/python/apache_beam/internal/metrics/metric.py

```diff
@@ -0,0 +1,138 @@
+# pytype: skip-file
+# mypy: disallow-untyped-defs
+
+from __future__ import absolute_import
+
+import datetime
+import logging
+import threading
+import time
+from builtins import object
+from typing import TYPE_CHECKING
+from typing import Dict
+from typing import Optional
+from typing import Type
+from typing import Union
+
+from apache_beam.internal.metrics.cells import HistogramCellFactory
+from apache_beam.metrics.execution import MetricUpdater
+from apache_beam.metrics.metric import Metrics as UserMetrics
+from apache_beam.metrics.metricbase import Histogram
+from apache_beam.metrics.metricbase import MetricName
+
+if TYPE_CHECKING:
+  from apache_beam.metrics.cells import MetricCell
+  from apache_beam.metrics.cells import MetricCellFactory
+  from apache_beam.utils.histogram import BucketType
+
+__all__ = ['Metrics']
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class Metrics(object):
+  @staticmethod
+  def histogram(namespace, name, bucket_type, logger=None):
+    # type: (Union[Type, str], str, BucketType, Optional[MetricLogger]) -> Metrics.DelegatingHistogram
+
+    """Obtains or creates a Histogram metric.
+
+    Args:
+      namespace: A class or string that gives the namespace to a metric
+      name: A string that gives a unique name to a metric
+      bucket_type: A type of bucket used in a histogram. A subclass of
+        apache_beam.utils.histogram.BucketType
+      logger: MetricLogger for logging locally aggregated metric
+
+    Returns:
+      A Histogram object.
+    """
+    namespace = UserMetrics.get_namespace(namespace)
+    return Metrics.DelegatingHistogram(
+        MetricName(namespace, name), bucket_type, logger)
+
+  class DelegatingHistogram(Histogram):
+    """Metrics Histogram that Delegates functionality to MetricsEnvironment."""
+    def __init__(self, metric_name, bucket_type, logger):
+      # type: (MetricName, BucketType, Optional[MetricLogger]) -> None
+      super(Metrics.DelegatingHistogram, self).__init__(metric_name)
+      self.metric_name = metric_name
+      self.cell_type = HistogramCellFactory(bucket_type)
+      self.logger = logger
+      self.updater = MetricUpdater(self.cell_type, self.metric_name)
+
+    def update(self, value):
+      # type: (object) -> None
+      self.updater(value)
+      if self.logger:
+        self.logger.update(self.cell_type, self.metric_name, value)
+
+
+class MetricLogger(object):
+  """Simple object to locally aggregate and log metrics.
+
+  This class is experimental. No backwards-compatibility guarantees.
+  """
+  def __init__(self):
+    # type: () -> None
+    self._metric = dict()  # type: Dict[MetricName, MetricCell]
+    self._lock = threading.Lock()
+    self._last_logging_millis = int(time.time() * 1000)
+    self.minimum_logging_frequency_msec = 180000
+
+  def update(self, cell_type, metric_name, value):
+    # type: (Union[Type[MetricCell], MetricCellFactory], MetricName, object) -> None
+    cell = self._get_metric_cell(cell_type, metric_name)
```

Comment:
So justt o confirm, every time the metric is updated, regardless of which environment is set. Thee same cell will be obtained and updated here. Which aggregates everything in the process together.

Reply:
True. The reason was for streaming pipelines the bundle size is too small so that per bundle metrics are not really helpful especially for statistics like histogram percentiles. `MetricLogger` could be considered as a process-wide shared metric container.